### PR TITLE
Initial commit for parsing rtti structures

### PR DESCRIPTION
**Pull request by PankajKataria**
Added data structures used for MVSC compiler RTTI, added some initial functions for parsing RTTI structures.

### CODE REVIEW

File: libr/core/anal_vt.c [hunk 1]
```diff
@@ -183,3 +183,41 @@ static void r_core_anal_list_vtables_all(void *core) {
 	r_list_free (vtables);
 }
 
+rtti_struct* get_rtti_data (RCore *core, ut64 atAddress) {
```
Comment (radare) [inline libr/core/anal_vt.c]: this must be static

File: libr/core/anal_vt.c [hunk 2]
```diff
@@ -183,3 +183,41 @@ static void r_core_anal_list_vtables_all(void *core) {
...
+	ut64 bits = r_config_get_i (core->config, "asm.bits");
+	int wordSize = bits / 8;
+	ut64 curRTTIBaseLocatorAddr = r_io_read_i (core->io, atAddress - wordSize, wordSize);
+	r_cons_printf ("trying to parse rtti at 0x%08"PFMT64x"\n", curRTTIBaseLocatorAddr);
+	return NULL;
+}
+
```
Comment (radare) [inline libr/core/anal_vt.c]: empty line

File: libr/core/anal_vt.c [hunk 3]
```diff
@@ -183,3 +183,41 @@ static void r_core_anal_list_vtables_all(void *core) {
...
+	return rtti_structures;
+}
+
+R_API void r_core_anal_print_rtti (void *core) {
+	RList* rtti_structures = r_core_anal_parse_rtti (core, false);
+	RListIter* RTTIIter;
+	rtti_struct* curRTTI;
+	if (rtti_structures) {
+		r_list_foreach (rtti_structures, RTTIIter, curRTTI) {
+			r_cons_printf ("Reached here\n");
```
Comment (radare) [inline libr/core/anal_vt.c]: use eprintf. this is a debug printf, right?

File: libr/core/anal_vt.c [hunk 4]
```diff
@@ -183,3 +183,41 @@ static void r_core_anal_list_vtables_all(void *core) {
 	r_list_free (vtables);
 }
 
+rtti_struct* get_rtti_data (RCore *core, ut64 atAddress) {
+	ut64 bits = r_config_get_i (core->config, "asm.bits");
+	int wordSize = bits / 8;
+	ut64 curRTTIBaseLocatorAddr = r_io_read_i (core->io, atAddress - wordSize, wordSize);
```
Comment (radare) [inline libr/core/anal_vt.c]: this var name is too long

File: libr/core/anal_vt.c [hunk 5]
```diff
@@ -183,3 +183,41 @@ static void r_core_anal_list_vtables_all(void *core) {
...
+	r_cons_printf ("trying to parse rtti at 0x%08"PFMT64x"\n", curRTTIBaseLocatorAddr);
```
Comment (radare) [inline libr/core/anal_vt.c]: start strings in uppercase

**Comment by radare** [comment]
Looks good to me. i'll merge as soon as you fix the comments. Can you submit some tests in r2r too?

**Comment by radare** [comment]
typedefs and static function names should be in camelcase

**Comment by radare** [comment]
RTTI is uppercase

**Comment by PankajKataria** [comment]
@radare yes sure, I have some more changes in mind and will submit the test after that, thanks for the review.

File: libr/core/anal_vt.c [hunk 6]
```diff
@@ -183,3 +183,40 @@ static void r_core_anal_list_vtables_all(void *core) {
...
+
+RList* r_core_anal_parse_rtti (void *core, bool printJson) {
+	RList* vtables = search_virtual_tables ((RCore *)core);
+	RListIter* vtableIter;
+	RList* rtti_structures = r_list_new();
+	vtable_info* table;
+
+	if (vtables) {
+		r_list_foreach (vtables, vtableIter, table) {
+			rtti_struct* current_rtti = get_rtti_data ((RCore *)core, table->saddr);
```
Comment (alvarofe) [inline libr/core/anal_vt.c]: This just return NULL, I assume is still incomplete.

File: libr/core/anal_vt.c [hunk 7]
```diff
@@ -183,3 +183,40 @@ static void r_core_anal_list_vtables_all(void *core) {
...
+	int wordSize = bits / 8;
+	ut64 BaseLocatorAddr = r_io_read_i (core->io, atAddress - wordSize, wordSize);
+	r_cons_printf ("Trying to parse rtti at 0x%08"PFMT64x"\n", BaseLocatorAddr);
+	return NULL;
+}
+
+RList* r_core_anal_parse_rtti (void *core, bool printJson) {
+	RList* vtables = search_virtual_tables ((RCore *)core);
+	RListIter* vtableIter;
+	RList* rtti_structures = r_list_new();
```
Comment (alvarofe) [inline libr/core/anal_vt.c]: use r_list_newf () to assign a free function otherwise there will be memory leaks.

File: libr/core/anal_vt.c [hunk 8]
```diff
@@ -183,3 +183,40 @@ static void r_core_anal_list_vtables_all(void *core) {
 	r_list_free (vtables);
 }
 
+static rtti_struct* get_rtti_data (RCore *core, ut64 atAddress) {
+	ut64 bits = r_config_get_i (core->config, "asm.bits");
+	int wordSize = bits / 8;
+	ut64 BaseLocatorAddr = r_io_read_i (core->io, atAddress - wordSize, wordSize);
+	r_cons_printf ("Trying to parse rtti at 0x%08"PFMT64x"\n", BaseLocatorAddr);
```
Comment (alvarofe) [inline libr/core/anal_vt.c]: I would add a message saying this is not done to avoid people complaining.

Comment (radare) [inline libr/core/anal_vt.c]: this seems like a debug printf, remove it or use eprintf

File: libr/core/anal_vt.c (hunk 7 above)
Comment (radare) [inline libr/core/anal_vt.c]: missing space

File: libr/core/anal_vt.c [hunk 9]
```diff
@@ -183,3 +183,40 @@ static void r_core_anal_list_vtables_all(void *core) {
...
+		}
+	}
+	r_list_free (vtables);
+	return rtti_structures;
+}
+
+R_API void r_core_anal_print_rtti (void *core) {
+	RList* rtti_structures = r_core_anal_parse_rtti (core, false);
+	RListIter* RTTIIter;
+	rtti_struct* curRTTI;
```
Comment (radare) [inline libr/core/anal_vt.c]: you can use shorter names here, RTTI is implicit, just 'cur'. also use r2-indent, that will put the * away from the typedef

File: libr/core/anal_vt.c [hunk 10]
```diff
@@ -183,3 +183,40 @@ static void r_core_anal_list_vtables_all(void *core) {
...
+	if (rtti_structures) {
+		r_list_foreach (rtti_structures, RTTIIter, curRTTI) {
+			eprintf ("Reached here\n");
```
Comment (radare) [inline libr/core/anal_vt.c]: left over? whats the point on iterating if you just print a debug message

File: libr/core/anal_vt.c [hunk 11]
```diff
@@ -183,3 +183,40 @@ static void r_core_anal_list_vtables_all(void *core) {
...
+			}
+		}
+	}
+	r_list_free (vtables);
+	return rtti_structures;
+}
+
+R_API void r_core_anal_print_rtti (void *core) {
+	RList* rtti_structures = r_core_anal_parse_rtti (core, false);
+	RListIter* RTTIIter;
```
Comment (radare) [inline libr/core/anal_vt.c]: just name this `iter`, like all the other RListIters in r2land

File: libr/core/cmd_anal.c [hunk 12]
```diff
@@ -5339,7 +5339,8 @@ static void r_anal_virtual_functions(void *core, const char* input) {
 			curArch = c->bin->cur->o->info->arch;
 		}
 	}
-	if (curArch && !strcmp (curArch, "x86")) {
+	if (curArch && !strcmp (curArch, "x86") && \
+			!strcmp (((RCore*) core)->bin->cur->o->info->rclass, "elf")) {
```
Comment (radare) [inline libr/core/cmd_anal.c]: i think this can produce a null deref

**Comment by radare** [comment]
ping

**Comment by radare** [comment]
please update the pr asap to merge. release is comming

**Comment by PankajKataria** [comment]
@radare updated

**Pull request merged**